# Empirical IO PS 2
Maximilian Huber

This code is stored at: https://github.com/MaximilianJHuber/NYU/blob/master/EmpIO/PS2.ipynb.
The notation follows along the lines of Berry, Levinsohn, Pakes (1995): http://www.tcd.ie/Economics/staff/ppwalsh/papers/BLP.pdf

In [1]:
using DataFrames
using GLM
using Optim
using LaTeXStrings

data = readtable("data_ps2.txt", header=false, separator=',')
rename!(data, names(data), [:car, :year, :firm, :price, :quantity, :weight, :hp, :ac, :nest3, :nest4]);

  likely near In[1]:6
  likely near In[1]:6

A car can change characteristics over the years:

In [2]:
data[data[:car] .== 91, :]

, use String instead.
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6

3×10 DataFrames.DataFrame
│ Row │ car │ year │ firm │ price │ quantity │ weight │ hp │ ac │ nest3 │
├─────┼─────┼──────┼──────┼───────┼──────────┼────────┼────┼────┼───────┤
│ 1   │ 91  │ 1990 │ 19   │ 5995  │ 52409.0  │ 1620   │ 55 │ 0  │ 1     │
│ 2   │ 91  │ 1991 │ 19   │ 6807  │ 55056.0  │ 1620   │ 55 │ 1  │ 1     │
│ 3   │ 91  │ 1992 │ 19   │ 8219  │ 66046.0  │ 1576   │ 55 │ 1  │ 1     │

│ Row │ nest4 │
├─────┼───────┤
│ 1   │ 1     │
│ 2   │ 1     │
│ 3   │ 1     │

 instead.
  likely near In[1]:6


Therefore, one can argue that product and time dimensions collapse into one. But for the time being lets treat good $j$ to be a _car_, time $t$ to be a _year_ and characteristics $x$ to be _weight_, _horse power_ and _air conditioning_ and finally the price $p$ be _price_.

## Part A: Logit

This part follows BLP (1995) section 6.1.

### 1
Agent $i$ derives utility from good $j$ at time $t$ in the following manner:
$$u_{ijt} = \delta_{jt}^* + \epsilon_{ijt} \quad \text{and} \quad \delta_{jt}^* = x'_{jt}\beta_t - \alpha_t p_{jt}+ \xi_{jt} \quad \forall t \in T$$
where $\epsilon_{ijt}$ is an i.i.d. extrem value.
This logit model has market shares:
$$s_{jt} = \frac{e^{\delta_{jt}^*}}{1 + \sum_{k=1}^J e^{\delta_{kt}^*}}$$
Taking logs yields:
$$\log s_{jt} - \log s_{0t} = \delta_{jt}^* = x'_{jt}\beta_t - \alpha_t p_{jt} + \xi_{jt} $$
where $\xi_{jt}$ is the unobservable good and time specific utility.

This equation will be estimated year-by-year.

### 2 
If I had chosen instead to pool the years the derivation would not change, but panel logit relies on the time-independence of $\xi_{jt}$, which is very implausible.

### 3 
Now I estimate the model with GMM using the BLP instrument for $p$:

#### Data Preparation
Market shares are calculated with an assumed market size of 100 million:

In [3]:
data[:share] = data[:quantity] / 1e8;

  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6

Instruments for the price are constructed by looking at the good's competitors, as defined by the goods produced by other firms, avaiable in the same year. I average of the characteristics of those competing goods.

I normalize weight, horse power and price BEFORE creating the instruments, to get the condition number of the matrices down:

In [4]:
data[:weight] = data[:weight] / mean(data[:weight])
data[:hp]     = data[:hp] / mean(data[:hp])
data[:price]  = data[:price] / mean(data[:price]);

String instead.
  likely near In[1]:6
  likely near In[1]:6

In [5]:
function competitor_mean_characteristic(good, char::Symbol)
    mean(data[(data[:firm] .!= good[:firm]) .* (data[:year] .== good[:year]), char]) #same year, different firm
end

 instead.
  likely near In[1]:6
  likely near In[1]:6

competitor_mean_characteristic (generic function with 1 method)

In [6]:
data[:comp_weight] = [competitor_mean_characteristic(good, :weight) for good in eachrow(data)]
data[:comp_hp]     = [competitor_mean_characteristic(good, :hp) for good in eachrow(data)]
data[:comp_ac]     = [competitor_mean_characteristic(good, :ac) for good in eachrow(data)];

 instead.
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6

And the left hand side of the model is:

In [7]:
data[:LHS] = zeros(size(data)[1])

for y in [1990, 1991, 1992]
    data[data[:year] .== y, :LHS] = 
        log(data[data[:year] .== y, :share]) - log(1 - sum(data[data[:year] .== y, :share]))
end

, use String instead.
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6
in #readtable#84 at C:\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\Users\Max\AppData\Local\JuliaPro-0.6.0.1\pkgs-0.6.0.1\v0.6\DataFrames\src\dataframe\io.jl
  likely near In[1]:6
in #readtable#84 at C:\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\Users\Max\AppData\Local\JuliaPro-0.6.0.1\pkgs-0.6.0.1\v0.6\DataFrames\src\dataframe\io.jl

In [8]:
head(data)


  likely near In[1]:6
  likely near In[1]:6

6×15 DataFrames.DataFrame
│ Row │ car │ year │ firm │ price    │ quantity │ weight   │ hp       │ ac │
├─────┼─────┼──────┼──────┼──────────┼──────────┼──────────┼──────────┼────┤
│ 1   │ 91  │ 1990 │ 19   │ 0.297023 │ 52409.0  │ 0.555996 │ 0.408123 │ 0  │
│ 2   │ 35  │ 1990 │ 7    │ 0.430795 │ 17122.0  │ 0.76192  │ 0.742041 │ 0  │
│ 3   │ 61  │ 1990 │ 16   │ 0.438227 │ 65590.0  │ 0.897487 │ 0.690099 │ 0  │
│ 4   │ 52  │ 1990 │ 10   │ 0.536029 │ 49877.0  │ 0.866255 │ 0.682678 │ 0  │
│ 5   │ 26  │ 1990 │ 4    │ 0.683724 │ 35944.0  │ 0.748878 │ 0.860768 │ 0  │
│ 6   │ 54  │ 1990 │ 11   │ 0.84202  │ 4640.0   │ 0.937642 │ 0.949813 │ 1  │

│ Row │ nest3 │ nest4 │ share      │ comp_weight │ comp_hp  │ comp_ac  │
├─────┼───────┼───────┼────────────┼─────────────┼──────────┼──────────┤
│ 1   │ 1     │ 1     │ 0.00052409 │ 0.990078    │ 0.97648  │ 0.427083 │
│ 2   │ 2     │ 2     │ 0.00017122 │ 1.00799     │ 1.00121  │ 0.45082  │
│ 3   │ 1     │ 1     │ 0.0006559  │ 1.00728     │ 1.01221  │ 0.504348 │
│ 4   │ 1     │ 1     │ 0.00049877 │ 1.00866     │ 1.00199  │ 0.472441 │
│ 5   │ 2     │ 3     │ 0.00035944 │ 1.00758     │ 0.996691 │ 0.468254 │
│ 6   │ 2     │ 2     │ 4.64e-5    │ 1.0036      │ 0.990654 │ 0.449612 │

│ Row │ LHS      │
├─────┼──────────┤
│ 1   │ -7.46288 │
│ 2   │ -8.58159 │
│ 3   │ -7.23853 │
│ 4   │ -7.5124  │
│ 5   │ -7.83999 │
│ 6   │ -9.88724 │

, use String

#### GMM Estimation
I estimate $\hat{\theta}$ using the following GMM procedure:
$$\max_\theta Q_n\big(\theta \big)$$
where $Q_n\big(\theta\big) = -\frac{1}{2}g_n\big(\theta\big)'\,\hat{W}\,g_n\big(\theta\big)$ and $g_n\big(\theta\big) = \frac{1}{n}\sum_{r=1}^R g\big(w_r; \theta\big)$. 

$g\big(w_r; \theta\big) = z_r \cdot (y_r - x_r \cdot \theta)$ is the residual $\xi_{jt}$ from the model above times the instruments calculated using $w_r$, a row of data.

The efficient GMM is with estimated weighting matrix $\hat{W}_A=\left(\frac{1}{n} \sum_{o=1}^n{g(w_o,\theta_{first})g(w_o,\theta_{first})'}\right)^{-1}$:

In [9]:
# w is data, with the left-hand-side as its first column, z are included regressors and excluded instruments
function gn(w, z, θ)

    1/size(w)[1] .* z' * (w[:,1] - w[:, 2:end] * θ)
    
end

#Wrapper creates a closure around the provided data set 
function Qn_wrapper(w, z)
    return θ -> 1/2 * (gn(w,z,θ)' * gn(w,z,θ))[1,1]
end

function Qn_wrapper(w, z, W)
    return θ -> 1/2 * (gn(w,z,θ)' * W * gn(w,z,θ))[1,1]
end


function g(w, z, θ)
    z * (w[1] - w[2:end]' * θ)
end

function What(w, z, θ)
    result = zeros(Float64, size(z)[2], size(z)[2])
    for i in 1:size(w)[1]
        result .+= g(w[i,:],z[i,:],θ) * g(w[i,:],z[i,:],θ)'
    end
    return inv(result/size(z)[1])
end

 instead.
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6
  likely near In[1]:6


What (generic function with 1 method)

In [10]:
table3 = DataFrame()
table3[:year] = [1990, 1991, 1992]

table3 = hcat(table3, convert(DataFrame, hcat(
    [begin
        w = convert(Array{Float64}, data[data[:year] .== year, [:LHS, :weight, :hp, :ac, :price]])
        z = convert(Array{Float64}, data[data[:year] .== year, [:weight, :hp, :ac, :comp_weight, :comp_hp, :comp_ac]])

        Qn = Qn_wrapper(w, z)

        optres = optimize(Qn, zeros(4), BFGS())
        θfirst = optres.minimizer

        Qn = Qn_wrapper(w, z, What(w, z, θfirst))
        optres = optimize(Qn, θfirst, BFGS())
        optres.minimizer
    end for year in [1990, 1991, 1992]]...)')
)

rename!(table3, names(table3)[2:end], [:βw, :βhp, :βac, :α])
table3[:α] = - table3[:α]

table3

  likely near In[1]:6
in builddf at C:\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\Users\Max\AppData\Local\JuliaPro-0.6.0.1\pkgs-0.6.0.1\v0.6\DataFrames\src\dataframe\io.jl


3×5 DataFrames.DataFrame
│ Row │ year │ βw        │ βhp      │ βac       │ α        │
├─────┼──────┼───────────┼──────────┼───────────┼──────────┤
│ 1   │ 1990 │ -0.147736 │ -19.2233 │ -3.19134  │ -13.7527 │
│ 2   │ 1991 │ -0.951378 │ -15.0633 │ -1.0394   │ -8.67934 │
│ 3   │ 1992 │ -1.45233  │ -13.1741 │ -0.962987 │ -6.83927 │

### 4
Cross-price elasticities are constant over all goods, which is rather unrealistic. For example, increasing the price of a Aston Martin DB11 would shift market share equally to the Ford Prius and the Ferrari 911. 

Similarily, the own-price elasticity is constant, but in reality some cars will have higher and some lower.
## Part B: Nested Logit
### 5
Using Berry (1994) notation:

$$\log s_{jt} - \log s_{0t} = \frac{\delta_{jt}^*}{1-\sigma} - \sigma \log{D_g}$$
where $D_g=\sum_{j \in g} e^{\delta_{jt}^*/(1-\sigma)}$. That yields:

$$\log s_{jt} - \log s_{0t} = x'_{jt}\beta_t - \alpha_t p_{jt} + \sigma \log s_{j|g} + \xi_{jt}  $$
### 6
Let me calculate the within group shares:

In [11]:
for nest_structure in [:nest3, :nest4]
    # initialize a column of zeros
    data[Symbol(string("share_", string(nest_structure)))] = zeros(size(data)[1])

    for y in [1990, 1991, 1992]
        for nest in 1:maximum(data[nest_structure])
            data[(data[:year] .== y) .* (data[nest_structure] .== nest), Symbol(string("share_", string(nest_structure)))] = 
                log(data[(data[:year] .== y) .* (data[nest_structure] .== nest), :share]) - 
                log(sum(data[(data[:year] .== y) .* (data[nest_structure] .== nest), :share]))
        end
    end
end

And the within group competitor's characteristics, which are being used as an instrument for the within shares:

In [12]:
function nest_competitor_mean_characteristic(good, char::Symbol, nest_structure::Symbol)
    mean(data[(data[nest_structure] .== good[nest_structure]) .* (data[:year] .== good[:year]) .* 
            (data[:car] .!= good[:car]), char]) #same year, same nest, different car
end

nest_competitor_mean_characteristic (generic function with 1 method)

In [13]:
for nest_structure in [:nest3, :nest4]

    data[Symbol(string("comp_weight_", string(nest_structure)))] = 
        [nest_competitor_mean_characteristic(good, :weight, nest_structure) for good in eachrow(data)]

    data[Symbol(string("comp_hp_", string(nest_structure)))] = 
        [nest_competitor_mean_characteristic(good, :hp, nest_structure) for good in eachrow(data)]

    data[Symbol(string("comp_ac_", string(nest_structure)))] = 
        [nest_competitor_mean_characteristic(good, :ac, nest_structure) for good in eachrow(data)];
    
end

In [14]:
head(data)

6×23 DataFrames.DataFrame
│ Row │ car │ year │ firm │ price    │ quantity │ weight   │ hp       │ ac │
├─────┼─────┼──────┼──────┼──────────┼──────────┼──────────┼──────────┼────┤
│ 1   │ 91  │ 1990 │ 19   │ 0.297023 │ 52409.0  │ 0.555996 │ 0.408123 │ 0  │
│ 2   │ 35  │ 1990 │ 7    │ 0.430795 │ 17122.0  │ 0.76192  │ 0.742041 │ 0  │
│ 3   │ 61  │ 1990 │ 16   │ 0.438227 │ 65590.0  │ 0.897487 │ 0.690099 │ 0  │
│ 4   │ 52  │ 1990 │ 10   │ 0.536029 │ 49877.0  │ 0.866255 │ 0.682678 │ 0  │
│ 5   │ 26  │ 1990 │ 4    │ 0.683724 │ 35944.0  │ 0.748878 │ 0.860768 │ 0  │
│ 6   │ 54  │ 1990 │ 11   │ 0.84202  │ 4640.0   │ 0.937642 │ 0.949813 │ 1  │

│ Row │ nest3 │ nest4 │ share      │ comp_weight │ comp_hp  │ comp_ac  │
├─────┼───────┼───────┼────────────┼─────────────┼──────────┼──────────┤
│ 1   │ 1     │ 1     │ 0.00052409 │ 0.990078    │ 0.97648  │ 0.427083 │
│ 2   │ 2     │ 2     │ 0.00017122 │ 1.00799     │ 1.00121  │ 0.45082  │
│ 3   │ 1     │ 1     │ 0.0006559  │ 1.00728     │ 1.01221  │ 0.504348 │
│ 4   │ 1     │ 1     │ 0.00049877 │ 1.00866     │ 1.00199  │ 0.472441 │
│ 5   │ 2     │ 3     │ 0.00035944 │ 1.00758     │ 0.996691 │ 0.468254 │
│ 6   │ 2     │ 2     │ 4.64e-5    │ 1.0036      │ 0.990654 │ 0.449612 │

│ Row │ LHS      │ share_nest3 │ share_nest4 │ comp_weight_nest3 │
├─────┼──────────┼─────────────┼─────────────┼───────────────────┤
│ 1   │ -7.46288 │ -4.30458    │ -3.68471    │ 0.9529            │
│ 2   │ -8.58159 │ -5.57302    │ -5.76142    │ 1.00937           │
│ 3   │ -7.23853 │ -4.08023    │ -3.46037    │ 0.945476          │
│ 4   │ -7.5124  │ -4.3541     │ -3.73423    │ 0.946155          │
│ 5   │ -7.83999 │ -4.83142    │ -3.3802     │ 1.00956           │
│ 6   │ -9.88724 │ -6.87867    │ -7.06707    │ 1.00683           │

│ Row │ comp_hp_nest3 │ comp_ac_nest3 │ comp_weight_nest4 │ comp_hp_nest4 │
├─────┼───────────────┼───────────────┼───────────────────┼───────────────┤
│ 1   │ 0.751882      │ 0.26087       │ 0.911598          │ 0.686388      │
│ 2   │ 1.02735       │ 0.492754      │ 0.990552          │ 0.927349      │
│ 3   │ 0.745752      │ 0.26087       │ 0.898464          │ 0.675543      │
│ 4   │ 0.745913      │ 0.26087       │ 0.899665          │ 0.675829      │
│ 5   │ 1.02563       │ 0.492754      │ 1.13845           │ 1.34546       │
│ 6   │ 1.02434       │ 0.478261      │ 0.988145          │ 0.924502      │

│ Row │ comp_ac_nest4 │
├─────┼───────────────┤
│ 1   │ 0.115385      │
│ 2   │ 0.424658      │
│ 3   │ 0.115385      │
│ 4   │ 0.115385      │
│ 5   │ 0.863636      │
│ 6   │ 0.410959      │

#### 6.a.

In [15]:
nest_structure = :nest3

table6 = DataFrame()
table6[:year] = [1990, 1991, 1992, 1990, 1991, 1992]
table6[:nesting] = [3, 3, 3, 4, 4, 4]

table6 = hcat(table6, convert(DataFrame, hcat(
    [begin
        # do 2SLS:
        X = convert(Array{Float64}, data[data[:year] .== year, 
                [:weight, :hp, :ac, :price, Symbol(string("share_", string(nest_structure)))]])
        Z = convert(Array{Float64}, data[data[:year] .== year, 
                [:weight, :hp, :ac, :comp_weight, :comp_hp, :comp_ac, Symbol(string("comp_weight_", string(nest_structure))), 
                Symbol(string("comp_hp_", string(nest_structure))), Symbol(string("comp_ac_", string(nest_structure)))]])

        y = convert(Array{Float64}, data[data[:year] .== year, [:LHS]])

        θ2SLS = ((X'Z * (Z'Z)^(-1) * Z'X) \ (X'Z * (Z'Z)^(-1) * Z'y))[:,1]

        # so efficient GMM starting from 2SLS:
        w = hcat(data[data[:year] .== year, :LHS], X)
        z = Z

        Qn = Qn_wrapper(w, z)

        optres = optimize(Qn, θ2SLS, BFGS())
        θfirst = optres.minimizer

        Qn = Qn_wrapper(w, z, What(w, z, θfirst))
        optres = optimize(Qn, θfirst, BFGS())
        optres.minimizer
    end for nest_structure in [:nest3, :nest4] for year in [1990, 1991, 1992]] ...)')
)

rename!(table6, names(table6)[2:end], [:nest, :βw, :βhp, :βac, :α, :σ])
table6[:α] = - table6[:α]

table6

6×7 DataFrames.DataFrame
│ Row │ year │ nest │ βw        │ βhp      │ βac         │ α         │
├─────┼──────┼──────┼───────────┼──────────┼─────────────┼───────────┤
│ 1   │ 1990 │ 3    │ -1.27194  │ -4.37998 │ 0.0507618   │ -2.29995  │
│ 2   │ 1991 │ 3    │ -0.830164 │ -2.02383 │ -0.00395255 │ -0.862333 │
│ 3   │ 1992 │ 3    │ -1.47325  │ -2.53423 │ -0.185029   │ -1.2094   │
│ 4   │ 1990 │ 4    │ -0.141635 │ -7.54913 │ -0.159515   │ -3.85128  │
│ 5   │ 1991 │ 4    │ -1.24604  │ -5.95231 │ 0.198332    │ -2.87299  │
│ 6   │ 1992 │ 4    │ 2.03069   │ -7.25268 │ -0.448861   │ -2.34421  │

│ Row │ σ        │
├─────┼──────────┤
│ 1   │ 0.969095 │
│ 2   │ 1.25584  │
│ 3   │ 1.06647  │
│ 4   │ 0.85888  │
│ 5   │ 0.786682 │
│ 6   │ 1.10408  │

The estimates for the price coefficients are higher!

#### 6.b.
The conditional error must be time-independent. This is very implausible with real data, since brands have autocorrelation in how much they are liked by buyers.

#### 6.c.
$\sigma$ is often close to one, but in 1991 the two nesting structures give very different estimates for $\sigma$. With 4 nests the $\sigma=0.79$ indicates there is much correlation within the nest. With 3 nests the $\sigma=1.26$ might not be consistent with random utility maximization! 
#### 6.d.
The $\alpha$ also varies a lot over the three years! 
### 7
#### 7.a.
Lets have a look at price elasticities in the logit and the nested logit model:

|          |   Standard Logit   | Nested Logit  |
|:-:|:-:|:-:|
| $\frac{\partial s_{jt}}{\partial p_{jt}}$  | $-\alpha s_{jt}\big(1-s_{jt}\big)$  | $-\alpha s_{jt}\big(\frac{1}{1-\sigma}-\frac{\sigma}{1-\sigma}s_{jt|g}-s_{jt}\big)$  |
|  $\frac{\partial s_{jt}}{\partial p_{kt}}$ | $\alpha s_{jt}s_{kt}$  | $\quad\quad\begin{cases}
\alpha s_{kt}\big(\frac{\sigma}{1-\sigma}s_{jt|g}+s_{jt}\big) & \text{same}\\
\alpha s_{jt}s_{kt} & \text{different}
\end{cases}$  |

The nested logit allows for a more flexible substitution pattern!
#### 7.b.
One could grow a dicision tree (or use some other machine learning methods, "nearest neighbor", etc.) to flexibly use the existing data (characteristics only, no market shares) to create nests. As validation I would check whether the catagories resemble something like: hypercar/supercar, sports car, SUV, hatchback, ...
#### 7.c.
We need to define the nesting structure a priori and must not make any mistakes. It is clear that we would need multiple layers of nesting to get a realistic classification, but then it is hard to justify why to split first according to one criterium instead of another.

### 8
#### 8.a.
Probit would allow for a very flexible substitution pattern, but with more than 120 cars per year the covariance matrix would get very big!
#### 8.b.
The pure characteristics model could not reflect the (unobserved) emotional value of people buying an Alfa Romeo Giulia Quadrifoglio although it is a bad car for the price.

## Part C: Mixed Logit
### 9
Now:
$$u_{ijt} = x'_{jt}\beta_t - \alpha_{i} p_{jt}+ \xi_{jt} + \epsilon_{ijt} \quad \forall t \in T$$
where $\alpha_i=\frac{1}{y_i}$ and $y_i \sim \mathcal{LN}$, the log-normal distribution.

Compared to BLP (1995) this is different in the sense the my $\delta$ is now $x'_{jt}\beta_t - \mu_\alpha p_{jt} + \xi_{jt}$, the $\alpha$ is integrated out and its distributional parameters are estimated.

The mixed logit shares are in this setting:
$$s_{jt}=\int_{0}^{\infty}\frac{exp\big(x'_{jt}\beta_{t}-(\mu_{\alpha}-\mu_{\alpha}+\frac{1}{y_{i}})p_{jt}+\xi_{jt}\big)}{1+\sum_{k}exp\big(x'_{jt}\beta_{t}-\frac{1}{y_{i}}p_{jt}+\xi_{jt}\big)}\frac{1}{y_{i}\sigma_{y}\sqrt{2\pi}}\exp\Big(-\frac{(\log y_{i}-\mu_{y})^{2}}{2\sigma_{y}^{2}}\Big)dy_{i}=
$$

$$\int_{0}^{\infty}\frac{exp\big(\delta_{jt}+(\mu_{\alpha}-\frac{1}{y_{i}})p_{jt}\big)}{1+\sum_{k}exp\big(\delta_{jt}+(\mu_{\alpha}-\frac{1}{y_{i}})p_{jt}\big)}\frac{1}{y_{i}\sigma_{y}\sqrt{2\pi}}\exp\Big(-\frac{(\log y_{i}-\mu_{y})^{2}}{2\sigma_{y}^{2}}\Big)dy_{i}=
$$

$$\int_{0}^{\infty}\frac{exp\big(\delta_{jt}+(\mu_{\alpha}-\frac{1}{\exp(e_{i})})p_{jt}\big)}{1+\sum_{k}exp\big(\delta_{kt}+(\mu_{\alpha}-\frac{1}{\exp(e_{i})})p_{kt}\big)}\frac{1}{\exp(e_{i})\sigma_{y}\sqrt{2\pi}}\exp\Big(-\frac{(e_{i}-\mu_{y})^{2}}{2\sigma_{y}^{2}}\Big)\exp(e_{i})de_{i}=
$$

$$\int_{0}^{\infty}\frac{exp\Big(\delta_{jt}+(\mu_{\alpha}-\exp(-e_{i}))p_{jt}\Big)}{1+\sum_{k}exp\Big(\delta_{kt}+(\mu_{\alpha}-\exp(-e_{i}))p_{kt}\Big)}\frac{1}{\sigma_{y}\sqrt{2\pi}}\exp\Big(-\frac{(e_{i}-\mu_{y})^{2}}{2\sigma_{y}^{2}}\Big)de_{i}=
$$

$$\int_{0}^{\infty}\underset{h(e_{i};\delta_{jt},\,p_{jt})}{\underbrace{\frac{1}{\sqrt{\pi}}\frac{exp\Big(\delta_{jt}+\Big(\exp(-\mu_{y}+\frac{1}{2}\sigma_{y}^{2})-\exp(-\mu_{y}-\sqrt{2}\sigma_{y}e_{i})\Big)\,p_{jt}\Big)}{1+\sum_{k}exp\Big(\delta_{kt}+\Big(\exp(-\mu_{y}+\frac{1}{2}\sigma_{y}^{2})-\exp(-\mu_{y}-\sqrt{2}\sigma_{y}e_{i})\Big)\,p_{kt}\Big)}}}\phi(e_{i})de_{i}\approx
$$

$$\approx\sum_{l=1}^n q_l \, h\Big(e_{l};\delta_{jt},\,p_{jt}\Big)$$
where $q_l$ and $e_l$ are the Gauss-Hermite weights and points.

In [21]:
using FastGaussQuadrature

function gn_BLP(w, z, θ)
    1/size(w)[1] .* z' * (w[:,1] .- w[:, 2:4] * θ[1:3] .+ θ[4] .* w[:, 5])
end

#Wrapper creates a closure around the provided data set 
function Qn_wrapper_BLP(w, z)
    
    δ = zeros(size(w)[1])
    (e, q) = gausshermite(30)
    p = w[:, 5] #price
    s = w[:, 1] #share
    
    
    return θ -> begin #w, hp, ac, mu, sigma
        μ_y = θ[4]
        σ_y = θ[5]
        
        h(e) = 1/(√π) .* exp.(δ .+ (exp(-μ_y + 1/2 * σ_y^2) - exp(-μ_y - (√2) * σ_y * e)) .* p) / 
            (1 + sum(exp.(δ .+ (exp(-μ_y + 1/2 * σ_y^2) - exp(-μ_y - (√2) * σ_y * e)) .* p)))
        
        error = 1
        
        while error > 1e-13
            δold = copy(δ)
            δ = δold .+ log.(s) .- log.(hcat(h.(e)...) * q)
            error = maximum(abs.(δ - δold))
        end

        gvec = gn_BLP(hcat(δ, w[:, 2:5]), z, θ) 
        1/2 * (gvec' * gvec)[1,1]
    end
end

Qn_wrapper_BLP (generic function with 1 method)

In [26]:
year = 1990

w = convert(Array{Float64}, data[data[:year] .== year, [:share, :weight, :hp, :ac, :price]])
z = convert(Array{Float64}, data[data[:year] .== year, [:weight, :hp, :ac, :comp_weight, :comp_hp, :comp_ac]])

Qn = Qn_wrapper_BLP(w, z)

optres = optimize(Qn, vcat(zeros(4), 1), NelderMead())

Results of Optimization Algorithm
 * Algorithm: Nelder-Mead
 * Starting Point: [0.0,0.0,0.0,0.0,1.0]
 * Minimizer: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimum: 1.126676e-02
 * Iterations: 1000
 * Convergence: false
   *  √(Σ(yᵢ-ȳ)²)/n < 1.0e-08: false
   * Reached Maximum Number of Iterations: true
 * Objective Function Calls: 1403

In [27]:
optres = optimize(Qn, optres.minimizer, NelderMead())

Results of Optimization Algorithm
 * Algorithm: Nelder-Mead
 * Starting Point: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimizer: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimum: 1.126676e-02
 * Iterations: 1000
 * Convergence: false
   *  √(Σ(yᵢ-ȳ)²)/n < 1.0e-08: false
   * Reached Maximum Number of Iterations: true
 * Objective Function Calls: 1007

In [28]:
optres = optimize(Qn, optres.minimizer, NelderMead())

Results of Optimization Algorithm
 * Algorithm: Nelder-Mead
 * Starting Point: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimizer: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimum: NaN
 * Iterations: 1000
 * Convergence: false
   *  √(Σ(yᵢ-ȳ)²)/n < 1.0e-08: false
   * Reached Maximum Number of Iterations: true
 * Objective Function Calls: 1007

In [29]:
optres = optimize(Qn, optres.minimizer, NelderMead())

Results of Optimization Algorithm
 * Algorithm: Nelder-Mead
 * Starting Point: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimizer: [-5.691735033022672,-5.0549542860660175, ...]
 * Minimum: NaN
 * Iterations: 1000
 * Convergence: false
   *  √(Σ(yᵢ-ȳ)²)/n < 1.0e-08: false
   * Reached Maximum Number of Iterations: true
 * Objective Function Calls: 1007

In [ ]:
optres = optimize(Qn, optres.minimizer, NelderMead(), Optim.OptimizationOptions(iterations=100000))

Stacktrace:
 [1] depwarn(::